pkg/opensearch/opensearch.go

```diff
@@ -64,11 +65,53 @@ func (ds *OpenSearchDatasource) QueryData(ctx context.Context, req *backend.Quer
 		return nil, err
 	}
 
+	err = handleServiceMapPrefetch(ctx, osClient, req)
+	if err != nil {
+		return nil, err
+	}
+
 	query := newQueryRequest(osClient, req.Queries, req.PluginContext.DataSourceInstanceSettings, intervalCalculator)
 	response, err := wrapError(query.execute(ctx))
 	return response, err
 }
 
+// handleServiceMapPrefetch inspects the given request, and, if it wants a serviceMap, creates and
+// calls the Prefetch query to get the services and operations lists that are required for
+// the associated Stats query. It then adds these parameters to the originating query so
+// the Stats query can be created later.
+func handleServiceMapPrefetch(ctx context.Context, osClient client.Client, req *backend.QueryDataRequest) error {
+	var serviceMapQuery backend.DataQuery
+	var serviceMapQueryIndex int
+	for i, query := range req.Queries {
+		model, _ := simplejson.NewJson(query.JSON)
+		queryType := model.Get("queryType").MustString()
+		luceneQueryType := model.Get("luceneQueryType").MustString()
+		serviceMapRequested := model.Get("serviceMap").MustBool(false)
+		if queryType == Lucene && luceneQueryType == "Traces" && serviceMapRequested {
+			serviceMapQueryIndex = i
+			serviceMapQuery = createServiceMapPrefetchQuery(query)
+			break
+		}
+	}
+	if serviceMapQuery.JSON != nil {
+		query := newQueryRequest(osClient, []backend.DataQuery{serviceMapQuery}, req.PluginContext.DataSourceInstanceSettings, intervalCalculator)
+		response, _ := wrapError(query.execute(ctx))
```

Comment:
if the prefetch errors, do we return the response as an error? It looks like wrapError doesn't always send back the response? I am probably misunderstanding but it looks like we swallow the error here?

Reply:
Yeah, no, you're not misunderstanding 😅 I think we just missed this. It's handled now.

Reply:
We kinda still swallow the error tbh. As @sarahzinger mentioned, if we don't return the error in the Response field, it won't be shown to the end user. We started by writing the wrapError function to make it possible to return custom errors, but from tinkering with it a little bit this morning, I don't see a simple and pretty way to do it for the prefetch. Since this is the same problem we have with all the other query types (they all show "downstream plugin error") I would suggest we add this as an improvement later, if you agree?